packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1827,9 +1827,59 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
   return root;
 }
 
-export function warnIfNotCurrentlyBatchingInDev(fiber: Fiber): void {
+// in a test-like environment, we want to warn if dispatchAction() is
+// called outside of a TestUtils.act(...)/batchedUpdates/render call.
+// so we have a a step counter for when we descend/ascend from
+// actedUpdates() calls, and test on it for when to warn
+let actingUpdatesScopeDepth = 0;
+
+export function actedUpdates(callback: () => void | Promise<void>) {
+  let previousActingUpdatesScopeDepth;
   if (__DEV__) {
-    if (isRendering === false && isBatchingUpdates === false) {
+    previousActingUpdatesScopeDepth = actingUpdatesScopeDepth;
+    actingUpdatesScopeDepth++;
+  }
+
+  const result = batchedUpdates(callback);
+  if (result && result.then) {
```

Comment:
We try to avoid `a && ...`, if you want to check `!== null && !== undefined` let's do that instead.

Reply:
https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberUnwindWork.js#L203-L205